### PR TITLE
Add SSM build and speed up CI re-run

### PR DESCRIPTION
**Pull request by vastin**
**Description:**
Create and test ADOT Collector SSM package in CI workflow.
Make CI re-run faster by using cache hit flag to skip successful jobs.

**Link to tracking Issue:** <Issue number if applicable>

**Testing:** < Describe what testing was performed and which tests were added.>

**Documentation:** < Describe the documentation added.>

### CODE REVIEW

File: .github/workflows/CI.yml [hunk 1]
```diff
@@ -458,242 +547,326 @@ jobs:
...
     steps:
       - uses: actions/checkout@v2
 
+      - name: Cache if success
+        id: e2etest-ec2-2
+        uses: actions/cache@v2
+        with:
+          path: |
+            VERSION
+          key: e2etest-ec2-2-${{ github.run_id }}-${{ matrix.testcase }}-${{ matrix.testing_ami }}
```
Comment (pingleig) [inline .github/workflows/CI.yml:617]: So if the cache does not exists, it only got created when the job is success? 

https://docs.github.com/en/actions/guides/caching-dependencies-to-speed-up-workflows#example-using-the-cache-action

File: .github/workflows/CI.yml [hunk 2]
```diff
@@ -360,40 +434,55 @@ jobs:
...
           key: "cached_packages_${{ github.run_id }}"
 
+      - name: Create SSM package
+        if: steps.e2etest-release.outputs.cache-hit != 'true'
+        run: |
+          ssm_pkg_version=`cat build/packages/VERSION`
+          ssm_pkg_version="${ssm_pkg_version:1}"
+          python3 tools/ssm/ssm_manifest.py ${ssm_pkg_version}
+          aws s3 cp build/packages/ssm s3://aws-otel-collector-test/ssm --recursive
+          bash tools/ssm/ssm_create.sh testAWSDistroOTel-Collector ${ssm_pkg_version} aws-otel-collector-test/ssm
```
Comment (pingleig) [inline .github/workflows/CI.yml:474]: We might need to clean up the document though, the default quota for ssm accounts is not very high. Or we can remove the test ssm document if all the ec2 test has passed.

File: tools/ssm/linux-amd64-deb/install.sh [hunk 3]
```diff
@@ -0,0 +1,28 @@
+#!/bin/bash
```
Comment (pingleig) [inline tools/ssm/linux-amd64-deb/install.sh:1]: Use `#!/usr/bin/env bash` https://github.com/aws/amazon-cloudwatch-agent/blob/53f8c7cd6c2e1977bc45ea55f70298b0d7dce942/Tools/src/create_darwin.sh#L1 or `sh` https://github.com/aws/amazon-cloudwatch-agent/blob/53f8c7cd6c2e1977bc45ea55f70298b0d7dce942/packaging/dependencies/amazon-cloudwatch-agent-ctl#L1

Though I think all of them should work on most vms and bms.

File: .github/workflows/CI.yml (hunk 1 above)
Comment (vastin) [reply .github/workflows/CI.yml:617]: Yes, it is only created when job return success.

File: .github/workflows/CI.yml (hunk 2 above)
Comment (vastin) [reply .github/workflows/CI.yml:474]: Agree. I need to verify what happen when reaching SSM quota. We could add the clean up logic in  resource-clean.yml later.

File: tools/ssm/linux-amd64-deb/install.sh (hunk 3 above)
Comment (vastin) [reply tools/ssm/linux-amd64-deb/install.sh:1]: Agree. I would prefer to `#!/usr/bin/env bash` and verify in all supported AMIs.

File: tools/ssm/ssm_create.sh [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+set -eu
+
+PKG_NAME=$1
+REL_VERSION=$2
+S3_BUCKET=$3
+
+output=$(aws ssm list-documents --filter Key=Owner,Values=Self Key=DocumentType,Values=Package Key=Name,Values=${PKG_NAME})
+echo ${output}
+status=$(echo ${output} | python3 -c "import sys, json; print('empty') if len(json.load(sys.stdin)['DocumentIdentifiers']) == 0 else print('exist')")
```
Comment (mxiamxia) [inline tools/ssm/ssm_create.sh:23]: Q: do we need to check or pre-install if `python3 -c ` is not available on the target host?

File: tools/ssm/ssm_create.sh [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+if [[ ${status} == "empty" ]]
```
Comment (mxiamxia) [inline tools/ssm/ssm_create.sh:24]: Excuse me if has asked this before. For `aws ssm create-document` in CI workflow, does it just create/update the document in our Test account for integration tests. And the Prod version of document will be created in CD workflow?

File: tools/ssm/ssm_create.sh (hunk 4 above)
Comment (vastin) [reply tools/ssm/ssm_create.sh:23]: The python3 is listed in installed software of ubuntu-latest test runner as below.
https://github.com/actions/virtual-environments/blob/main/images/linux/Ubuntu2004-README.md#installed-software

File: tools/ssm/ssm_create.sh (hunk 5 above)
Comment (vastin) [reply tools/ssm/ssm_create.sh:24]: Exactly. The pro version of document will be created in CD workflow with same artifacts which passes CI workflow as release candidate.

**Review by mxiamxia — APPROVED** [review]
LGTM. Pls resolve the remaining comments before merging.

File: .github/workflows/CI.yml (hunk 2 above)
Comment (vastin) [reply .github/workflows/CI.yml:499]: The SSM quota issue will be addressed in a separate PR.

File: tools/ssm/linux-amd64-deb/install.sh (hunk 3 above)
Comment (vastin) [reply tools/ssm/linux-amd64-deb/install.sh]: Fixed

**Pull request merged**